File: py_capi/pytest/sat/SatSolver_test.py
#! /usr/bin/env python3

"""SatSolver のテストプログラム

:file: SatSolver_test.py
:author: Yusuke Matsunaga (松永 裕介)
:copyright: Copyright (C) 2022 Yusuke Matsunaga, All rights reserved.
"""

import pytest
from ymworks.sat import SatSolver, SatBool3


@pytest.fixture
def solver():
    return SatSolver()

def test_add_clause1(solver):
    v1 = solver.new_variable()
    v2 = solver.new_variable()
    v3 = solver.new_variable()

    solver.add_clause(v1, ~v2)
    solver.add_clause(~v1, v3)
    solver.add_clause(~v3)

    r = solver.solve()

    assert r == SatBool3.true

    model = solver.model()
    assert model[v1] == SatBool3.false
    assert model[v2] == SatBool3.false
    assert model[v3] == SatBool3.false

    assert model[v1] == solver.read_model(v1)
    assert model[v2] == solver.read_model(v2)
    assert model[v3] == solver.read_model(v3)

File: c++-srcs/sat/gtest/SatSolver_test.cc

/// @file sat_test.cc
/// @brief sat_test の実装ファイル
/// @author Yusuke Matsunaga (松永 裕介)
///
/// Copyright (C) 2023 Yusuke Matsunaga
/// All rights reserved.


#include "gtest/gtest.h"
#include "ym/SatSolver.h"


BEGIN_NAMESPACE_YM

TEST(SatSolverTest, test1)
{
  SatSolver solver{SatInitParam{"ymsat2"}};

  auto var0 = solver.new_variable(true);
  auto var1 = solver.new_variable(true);
  auto var2 = solver.new_variable(true);
  auto var3 = solver.new_variable(true);
  auto var4 = solver.new_variable(true);
  auto var5 = solver.new_variable(true);

  solver.add_buffgate(var1, var2);
  solver.add_orgate(var2, var3, var4);
  solver.add_notgate(var3, var0);
  solver.add_notgate(var4, var5);
  solver.add_notgate(var5, var0);

  auto assumptions = std::vector<SatLiteral>{~var2};

  auto ret = solver.solve(assumptions);

}

END_NAMESPACE_YM
